scintillometry/pulse_folding.py

```diff
@@ -0,0 +1,117 @@
+# Licensed under the GPLv3 - see LICENSE
+
+import operator
+import warnings
+
+import numpy as np
+import astropy.units as u
+import astropy.time as time
+from astropy.utils import lazyproperty
+from .base import Base, TaskBase
+
+
+__all__ = ['TimeFold',]
+
+#TODO add Fold base class for the future folding
+class TimeFold(TaskBase):
+    """ Fold pulse using a fixed time interval.
+
+    NOTE
+    ----
+    This method does not require the folding time to be the pulse period.
+    However, the time integration for each phase bin will not be uniform.
+
+    Parameters
+    ----------
+    ih : task or `baseband` stream reader
+        Input data stream, with time as the first axis.
+    phase_bin : int
+        Number of bins in one result phase period.
+    fold_time : float or `~astropy.units.Quantity`,
+        Time interval for folding into one pulse period(result sample).
+        Default unit is `second`.
+    phase_method : function
+        The method to compute pulse phases at a given time.
+        Note, the input time is in the format of '~astropy.time.Time' object.
+        The output phase should be an array.
+    samples_per_frame : int, optional
+        Number of request folded period. If not given, it assumes 1.
+    average : bool
+        If the output pulse profile averaged by the number of time integration
+        in each phase bin.
+    """
+
+    def __init__(self, ih, phase_bin, fold_time, phase_method,
+                 samples_per_frame=None, average=False):
+        self.ih = ih
+        self.phase_bin = phase_bin
+        if not hasattr(fold_time, 'unit'):
+            fold_time *= u.s
+        self.fold_time = fold_time
+        # Setup the number of returned samples
+        if samples_per_frame is None:
+            samples_per_frame = 1
+
+        self.phase_method = phase_method
+        self.average = average
+        super().__init__(ih, samples_per_frame=samples_per_frame,
+                         sample_rate=1./fold_time,
+                         shape=(samples_per_frame, phase_bin) + ih.shape[1:],
+                         dtype=ih.dtype)
+
+    def eval_phase(self, t):
+        """ Evaluate the pulse phase at a given time. """
+        return self.phase_method(t)
+
+    def _read_frame(self, frame_index):
+        # Move to the start position
+        result = np.zeros(self.shape, dtype=self.dtype)
+        counts = np.zeros(self.shape[0:2], dtype=np.int)
+        self.ih.seek(frame_index * self.samples_per_frame)
+        req_samples = int(self.samples_per_frame / self.sample_rate * \
+                      self.ih.sample_rate)
+        start_time = self.ih.time
+        data = self.ih.read(req_samples)
+        time_axis = start_time + np.arange(req_samples) / \
```

Comment:
It may not be a bad idea to move the actual reading of the `data` below the calculation of the phases, since then the times can be discarded, i.e.,
```
time_offset = np.arange(self._raw_samples_per_frame) / self.ih.sample_rate
phases = self.phase(self.ih.time + time_offset)
data = ...
```
(Perhaps even calculate time index first as well. A possible speed-up would be to special-case `samples_per_frame=1`, where the time index is not needed.)